boa/src/builtins/string/mod.rs

```diff
@@ -479,39 +479,90 @@ impl String {
             match replace_object {
                 Value::String(val) => {
                     // https://tc39.es/ecma262/#table-45
-                    let mut result = val.to_string();
-                    let re = Regex::new(r"\$(\d)").unwrap();
-
-                    if val.find("$$").is_some() {
-                        result = val.replace("$$", "$")
-                    }
-
-                    if val.find("$`").is_some() {
-                        let start_of_match = mat.start();
-                        let slice = &primitive_val[..start_of_match];
-                        result = val.replace("$`", slice);
-                    }
-
-                    if val.find("$'").is_some() {
-                        let end_of_match = mat.end();
-                        let slice = &primitive_val[end_of_match..];
-                        result = val.replace("$'", slice);
-                    }
-
-                    if val.find("$&").is_some() {
-                        // get matched value
-                        let matched = caps.get(0).expect("cannot get matched value");
-                        result = val.replace("$&", matched.as_str());
-                    }
-
-                    // Capture $1, $2, $3 etc
-                    if re.is_match(&result) {
-                        let mat_caps = re.captures(&result).unwrap();
-                        let group_str = mat_caps.get(1).unwrap().as_str();
-                        let group_int = group_str.parse::<usize>().unwrap();
-                        result = re
-                            .replace(result.as_str(), caps.get(group_int).unwrap().as_str())
-                            .to_string()
+                    let mut result = StdString::new();
+                    let mut chars = val.chars().peekable();
+
+                    let m = caps.len();
+
+                    while let Some(first) = chars.next() {
+                        if first == '$' {
+                            let second = chars.next();
+                            let second_is_digit = second.map(|ch| ch.is_digit(10)).unwrap_or(false);
+                            // we use peek so that it is still in the iterator if not used
+                            let third = if second_is_digit { chars.peek() } else { None };
+                            let third_is_digit = third.map(|ch| ch.is_digit(10)).unwrap_or(false);
+
+                            match (second, third) {
+                                (Some('$'), _) => {
+                                    // $$
+                                    result.push('$');
+                                }
+                                (Some('&'), _) => {
+                                    // $&
+                                    let matched = caps.get(0).expect("cannot get matched value");
```

Comment:
I would not panic here. We should return an error if this goes wrong, right? What does the spec say?

Reply:
By following the spec we wouldn't get this far if there is no match, https://tc39.es/ecma262/#sec-string.prototype.replace point 9 returns the original string if there is no match.
I'm gonna take a look into how to adapt this.

Reply:
https://github.com/RageKnify/boa/blob/a8222ced381fcab93a721323b94ff6b54427f061/boa/src/builtins/string/mod.rs#L471
It paniced earlier in the function when trying to get the match, I'll introduce a match statement on the option and a test covering this scenario.